Add simple read demo (#171)
NeurodataWithoutBorders/aqnwb: demo/inspect_electrical_series/README.md

# NWB Analysis Demo

This is a simple C++ demo project that shows how to read and analyze NWB (Neurodata Without Borders) files using the aqnwb library. The demo reads a NWB file and performs basic statistical analysis on the ElectricalSeries electrophysiology data.

## Prerequisites

- CMake (version 3.15 or higher)
- C++ compiler with C++17 support
- HDF5 library with C++ support
- Boost library
- aqnwb library (built from the parent project)

> ## Download an example NWB dataset that contains ElectricalSeries data
> 
> E.g. https://api.dandiarchive.org/api/assets/4c440a73-9250-45bb-a342-a7da4d01b2fd/download?content_disposition=inline to
> download a single file from DANDIset 000232 at https://dandiarchive.org/dandiset/000232/0.240510.2038
> 
> Alternatively, we can also run AqNWB test suite and use a suitable file from the tests, e.g,. 
> `build/dev/tests/data/createElectricalSeries.nwb`

Reviewer: ```suggestion
## Identify an example NWB dataset that contains ElectricalSeries data

1. For this example we will download a single file from [dandiset 000232](https://dandiarchive.org/dandiset/000232/0.240510.2038) using this [download link](https://api.dandiarchive.org/api/assets/4c440a73-9250-45bb-a342-a7da4d01b2fd/download?content_disposition=inline). https://api.dandiarchive.org/api/assets/4c440a73-9250-45bb-a342-a7da4d01b2fd/download?content_disposition=inline to
download a single file from DANDIset 000232 at https://dandiarchive.org/dandiset/000232/0.240510.2038
2. Place the file in the `aqnwb/demo/inspect_electrical_series` directory.
```

I think for the example readme it would be helpful to keep consistent using the DANDI file throughout.

## Building the Demo

1. Make sure you have built the main aqnwb project first 
   following the instructions in the AqNWB documentation

2. Create a build directory for the demo:

```bash

> # From the  demo/inspect_electrical_series directory

Reviewer: ```suggestion
cd demo/inspect_electrical_series
```

mkdir -p build
cd build
```

3. Configure and build the demo:

```bash
# Basic configuration
cmake ..

# Or with custom paths to dependencies
cmake .. \

>   -DAQNWB_DIR=../../build/dev \
>   -DHDF5_DIR=../../libs/hdf5_build \
>   -DBOOST_ROOT=../../libs/boost_1_82_0

Reviewer: ```suggestion
  -DAQNWB_DIR=../../../build/dev \
  -DHDF5_ROOT=../../libs/hdf5_build \
  -DBOOST_ROOT=../../libs/boost_1_82_0
```

You could update the path flag to HDF5_ROOT to keep consistent with the environment variable name used in the main installation instructions

Author: Done

# Build the project
make
```

## Running the Demo

After building, you can run the demo:

```bash
# From the demo/DANDISET_000232/build directory
cd bin

> ./nwb_analysis_demo ../../sub-EF0147_ses-20190204T144339_behavior+ecephys.nwb

Reviewer: When using the dandi demo file, I get this error:

```
Analyzing ElectricalSeries at: /processing/ecephys/LFP/ElectricalSeries_97
Number of time points: 7409
Number of channels: 64
Data unit: volts
Data description: no description
HDF5-DIAG: Error detected in HDF5 (1.14.3) thread 0:
  #000: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5D.c line 1061 in H5Dread(): can't synchronously read data
    major: Dataset
    minor: Read failed
  #001: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5D.c line 1008 in H5D__read_api_common(): can't read data
    major: Dataset
    minor: Read failed
  #002: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5VLcallback.c line 2092 in H5VL_dataset_read_direct(): dataset read failed
    major: Virtual Object Layer
    minor: Read failed
  #003: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5VLcallback.c line 2048 in H5VL__dataset_read(): dataset read failed
    major: Virtual Object Layer
    minor: Read failed
  #004: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5VLnative_dataset.c line 373 in H5VL__native_dataset_read(): can't read data
    major: Dataset
    minor: Read failed
  #005: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5Dio.c line 401 in H5D__read(): can't read data
    major: Dataset
    minor: Read failed
  #006: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5Dchunk.c line 2919 in H5D__chunk_read(): unable to read raw data chunk
    major: Low-level I/O
    minor: Read failed
  #007: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5Dchunk.c line 4548 in H5D__chunk_lock(): data pipeline read failed
    major: Dataset
    minor: Filter operation failed
  #008: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5Z.c line 1355 in H5Z_pipeline(): required filter 'deflate' is not registered
    major: Data filters
    minor: Read failed
  #009: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5PLint.c line 267 in H5PL_load(): can't find plugin. Check either HDF5_VOL_CONNECTOR, HDF5_PLUGIN_PATH, default location, or path set by H5PLxxx functions
    major: Plugin for dynamically loaded library
    minor: Object not found
  #010: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5PLpath.c line 804 in H5PL__find_plugin_in_path_table(): search in path /usr/local/hdf5/lib/plugin encountered an error
    major: Plugin for dynamically loaded library
    minor: Can't get value
  #011: /Users/smprince/Documents/projects/hdf5/hdf5-1.14.3/src/H5PLpath.c line 857 in H5PL__find_plugin_in_path(): can't open directory (/usr/local/hdf5/lib/plugin). Please verify its existence
    major: Plugin for dynamically loaded library
    minor: Can't open directory or file
libc++abi: terminating due to uncaught exception of type H5::DataSetIException
zsh: abort      ./nwb_analysis_demo ../../sub-EF0147_ses-20190204T144339_behavior+ecephys.nwb
```

It seems some filter plugins may be required to run this demo?

Author: Interesting. It worked for me. Not sure why this is failing.

Author: The `deflate` filter that the message mentions is `gzip`, which should be installed by default with HDF5.

Reviewer: I was using a version of HDF5 that was built from source, which doesn't seem to include the filter plugins by default. The other filter tests that you had added were also failing for similar reasons.

Updated to use the brew installation version instead and everything works!

# NWB Analysis Demo

> This is a simple C++ demo project that shows how to read and analyze NWB (Neurodata Without Borders) files using the aqnwb library. The demo reads the file `sub-EF0147_ses-20190204T144339_behavior+ecephys.nwb` and performs basic statistical analysis on the electrophysiology data.

Author: ```suggestion
This is a simple C++ demo project that shows how to read and analyze NWB (Neurodata Without Borders) files using the aqnwb library. The demo reads a NWB file and performs basic statistical analysis on the ElectricalSeries electrophysiology data.
```
